hubtools/src/lib.rs

```diff
@@ -652,23 +716,67 @@ impl RawHubrisArchive {
 
         let board = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("board")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
         let mut name = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("name")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
 
+        // The optional epoch argument and optional app.toml epoch
+        // parameter both default to zero and must agree with each other.
+        let epoch_arg: u32 = if let Some(arg) = epoch {
+            arg.parse::<u32>()
+                .map_err(|_| Error::InvalidEpoch(arg.to_string()))?
+        } else {
+            0u32
+        };
+
+        let epoch_toml: u32 = manifest
+            .as_table()
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .get("epoch")
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .as_integer()
+            .ok_or(Error::InvalidEpoch(format!("{}:{}", file!(), line!())))?
```

Comment:
I don't really like packing error tracebacks into the error here.  Having separate error types for `EpochNotAnInteger` versus `EpochOutOfRange` would handle these cases.

Reply:
...being reverted.